workflow/controller/operator.go

```diff
@@ -351,18 +351,23 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 		woc.log.WithError(err).Error("error in entry template execution")
 		// we wrap this error up to report a clear message
 		x := fmt.Errorf("error in entry template execution: %w", err)
+		shouldDeletePVCs := true
+
 		switch err {
 		case ErrDeadlineExceeded:
 			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", x.Error())
 		case ErrParallelismReached:
+			shouldDeletePVCs = false
 		default:
 			if !errorsutil.IsTransientErr(err) && !woc.wf.Status.Phase.Completed() && os.Getenv("BUBBLE_ENTRY_TEMPLATE_ERR") != "false" {
 				woc.markWorkflowError(ctx, x)
 			}
 		}
 		// Garbage collect PVCs if Entrypoint template execution returns error
-		if err := woc.deletePVCs(ctx); err != nil {
-			woc.log.WithError(err).Warn("failed to delete PVCs")
+		if shouldDeletePVCs {
+			if err := woc.deletePVCs(ctx); err != nil {
+				woc.log.WithError(err).Warn("failed to delete PVCs")
+			}
 		}
```

Comment:
I'm trying to understand. Is the difference here that some errors indicate that the Workflow should enter a failed state and other errors are a temporary problem that might be resolved at the next reconciliation? (which we could call "temporaryErrors" or something)

I see `markWorkflowError()` is not used in certain other cases, such as `ErrDeadlineExceeded`, transient error, etc. It seems to me that we shouldn't delete the PVCs in any cases where we're not considering the Workflow to be failed?

Reply:
If that is the case, it would be great to use the concept of "temporary error" generically and then perform whatever logic should happen when there's a non-temporary error (at this time maybe it's only the deletion of PVCs but in the future could be more)

Reply:
I'm having a hard time understanding the logic here myself to be honest. It might be that the PVC logic doesn't belong in that function at all.. 

I'm probably the wrong person to implement what you are proposing as I do not have a good overview of the code base..

Reply:
I think it looks like an okay place to put the logic since this is where the PVCs are being deleted. As far as I can tell, of all of the errors there, it's only where we `woc.markWorkflowError()` that there's a real error. Everywhere else is a temporary issue and the Workflow is left in its current state, to be reconciled later.

@jiachengxu I see you changed this code originally. Do you agree?

Reply:
@juliev0 I added the logic in multiple places in this PR: https://github.com/argoproj/argo-workflows/pull/10424 to fix the issue that PVC could not be GCed if Entrypoint template execution or Onexit template execution returns errors.
It was a bug reported here: https://github.com/argoproj/argo-workflows/issues/10408

Reply:
@jiachengxu do you agree that we should not delete the PVC in the case of certain transient/temporary errors? I am thinking we should only delete it if we're actually marking the Workflows as "Error" state.

Reply:
@juliev0 Yeah, I agree that for temporary errors we should not delete PVCs since that error could be resolved during the next reconciliation, and we only delete PVCs if the Workflow is in "real" Error state.

Reply:
@jiachengxu thanks for confirming! @abrabah do you have enough information to move forward?

Reply:
Thanks for the feedback, appreciate it 👍
I've pushed a commit where the PVC deletion logic has been moved to the default case in the switch statement